weaviate/collection/classes.py

```diff
@@ -88,69 +88,189 @@ class StopwordsPreset(str, Enum):
     EN = "en"
 
 
+class PQEncoderType(str, Enum):
+    KMEANS = "kmeans"
+    TILE = "tile"
+
+
+class PQEncoderDistribution(str, Enum):
+    LOG_NORMAL = "log-normal"
+    NORMAL = "normal"
+
+
 ModuleConfig = Dict[Vectorizer, Dict[str, Any]]
 
 
-class ConfigModel(BaseModel):
+class ConfigCreateModel(BaseModel):
     def to_dict(self):
         return self.model_dump(exclude_none=True)
 
 
-@dataclass
-class VectorIndexConfig(ConfigModel):
-    distance: VectorDistance = VectorDistance.COSINE
-    efConstruction: int = 128
-    maxConnections: int = 64
+class ConfigUpdateModel(BaseModel):
+    def merge_with_existing(self, schema: Dict[str, Any]) -> Dict[str, Any]:
+        for cls_field in self.model_fields:
+            val = getattr(self, cls_field)
+            if val is None:
+                continue
+            if isinstance(val, Enum):
+                schema[cls_field] = str(val.value)
+            elif isinstance(val, (int, float, bool, str, list)):
+                schema[cls_field] = val
+            else:
+                assert isinstance(val, ConfigUpdateModel)
+                schema[cls_field] = val.merge_with_existing(schema[cls_field])
+        return schema
+
+
+class PQEncoderConfigCreate(ConfigCreateModel):
+    type_: PQEncoderType = PQEncoderType.KMEANS
+    distribution: PQEncoderDistribution = PQEncoderDistribution.LOG_NORMAL
+
+    def merge_with_existing(self, schema: Dict[str, Any]) -> Dict[str, Any]:
+        """Must be done manually since Pydantic does not work well with type and type_.
+        Errors shadowing type occur if we want to use type as a field name.
+        """
+        if self.type_ is not None:
+            schema["type"] = str(self.type_.value)
+        if self.distribution is not None:
+            schema["distribution"] = str(self.distribution.value)
+        return schema
+
+
+class PQEncoderConfigUpdate(ConfigUpdateModel):
+    type_: Optional[PQEncoderType] = None
+    distribution: Optional[PQEncoderDistribution] = None
+
+    def merge_with_existing(self, schema: Dict[str, Any]) -> Dict[str, Any]:
+        """Must be done manually since Pydantic does not work well with type and type_.
+        Errors shadowing type occur if we want to use type as a field name.
+        """
+        if self.type_ is not None:
+            schema["type"] = str(self.type_.value)
+        if self.distribution is not None:
+            schema["distribution"] = str(self.distribution.value)
+        return schema
```

Comment:
combine with the one above?

Reply:
I'd hoped so but because the class vars have different types and defaults, the inheritance doesn't work nicely alongside the type hints and pydantic

The same error as documented in the docstring for `merge_with_existing` occurs if you try to overwrite a class var in a child with pydantic